honeybee/model.py

```diff
@@ -0,0 +1,480 @@
+# coding: utf-8
+"""Honeybee Model."""
+from .properties import ModelProperties
+from .room import Room
+from .face import Face
+from .shade import Shade
+from .aperture import Aperture
+from .door import Door
+from .boundarycondition import Outdoors, Surface
+from .facetype import AirWall
+from .typing import valid_string
+import honeybee.writer as writer
+
+from ladybug_geometry.geometry3d.face import Face3D
+
+
+class Model(object):
+    """A collection of Rooms, Faces, Shades, Apertures, and Doors representing a model.
+
+    Properties:
+        name
+        name_original
+        rooms
+        faces
+        shades
+        apertures
+        doors
+        has_ngon_apertures
+        had_ngon_doors
+    """
+    __slots__ = ('_name', '_name_original', '_rooms', '_faces', '_shades',
+                 '_apertures', '_doors', '_room_names', '_face_names', '_shade_names',
+                 '_aperture_names', '_door_names', '_properties')
+
+    def __init__(self, name, objects):
```

Comment:
@chriswmackey, We may want to discuss model over a call. There are a number of design decisions here that I want to know more why you made those choices.

1. Not sure if we need `x` and `x_names` properties. I would say a `x` method with a `names_only` argument looks like a cleaner choice to me.
2. Not a big fan of collecting objects as dictionary. I assume this decision was made to ease the process of returning a specific room, face or aperture but this can have some major performance downsides for us specially for large models. See this presentation for a reference: https://youtu.be/7eeEf_rAJds and since your next question is how is that possible when Python instances create \_\_dict__ unless using \_\_slots__? Well I don't know the exact answer but someone asks the same question at 33:40. 
2.1. if my assumption about access is correct using a quick search through a list or tuple just works as well. Specially if we have the separation based on type done in the model.

Reply:
@mostaphaRoudsari ,
It sounds like a call is in order. I really put the model class together relatively quickly, almost more as a record for all of the possible checks that we would like to build into it rather than anything that was optimized for accessing or storing other objects.

FYI, the reason for the `x_names` variables was to sense when there was a duplicate-named objects in the model since the dictionary key would be overwritten when this is the case.

I will take a look at that youtube video and I recognize that, if you are being mindful of the performance in this case after you have previously pushed to run a lot of geometry checks implicitly, then this dictionary situation must be a pretty big performance issue 🙂

Reply:
Thanks again for the reference video.  What the presenter said in regards to the question about dicted classes makes sense and I realize that there is a dark world of how the code is compiled that relies heavily on correctly interpreting people's 

I just wanted to outline the alternatives to the dictionary that seem worth considering here:

- two lists for each geometry object type - One with the names of the objects and another with the instances of the objects.  This might make lookups of objects by name a little slower since you will need to first find the index in the name list and then use that index to return the instance in the other list.  However, it would definitely make the size of the Model object in memory smaller since the two lists are pretty certainly less than the dictionary.  It also avoids the need for a separate lists of names from the one used to lookup objects, allowing use to check for duplicate names by just using the list with names.  Also, by using lists, it is relatively straightforward to add new objects to the model (we just append them to the two lists) and this also makes it easier to add all of the elements of parent objects to their respective lists (like all of the faces of a room to the "faces" list).  If optimized fast lookups is what we are willing to compromise on, this seems like the best solution.  And, thinking about it more, I am not sure that we will really use the lookups much beyond the check for adjacent objects with Surface boundary conditions.  Otherwise, people can easily access the faces of a room from the room object and there's no need to lookup the face on the model.

- namedtuple for each geometry object type - According to the talk, this does not have the same performance/memory issues as a dictionary and it will still allow for quick lookups.  However, I think the biggest drawback of this approach is that it's a tuple under the hood and not a list.  As a result, it becomes really intense to add new objects to the model or add all of the elements of a parent object since the tuple under the hood is immutable.  Personally, I don't think that having such optimized lookups is worth it here.

- relying on parent objects to store the only reference to their children - This will make lookups by name extremely painful.  However, it is the smallest size in memory.

Reply:
let's talk about all the comments about the model over the call. I actually think the third option is the best and don't see the look up to be as painful. That's how almost all d3js look-ups work for example.

Reply:
Ok, we can discuss once I finish the tests and merge it in.  And, just so that what we are merging something that we are actually considering, I switched over the model object to use the two lists (option 1) instead of the previous dictionary. On the call, we can go through option 3 and see how it affects each of the check methods.  As, I said before, I think option 3 is fine if we rarely see ourselves running the check methods.  I imagine that D3 might not have the same need to check things like the presence of adjacent objects in the model, duplicate names, non-adjacent air walls, and the other ~15 checks that are currently built into the Model object.